### ISSUE TITLE
Example integration with TorchElastic

### ISSUE THREAD

**Issue by twolffpiggott** [optuna/optuna#2735]
## Motivation

Optuna is an invaluable tool for model hyperparameter optimization, and many members of the community will likely be interested in using Optuna in conjunction with frameworks for fault-tolerant distributed training such as [TorchElastic](https://pytorch.org/elastic/0.2.3.dev0/index.html).

Currently Optuna supports and provides [examples](https://github.com/optuna/optuna-examples/blob/main/kubernetes/README.md) of distributed optimisation where *n independent jobs optimize n independent trials*. However, Optuna does not have a template (example optimization script) for usage with a Framework like TorchElastic, where Optuna is used in conjunction with distributed training, that is, *a group of n dependent elastic workers optimise a single model that corresponds to a single Optuna trial*.

Distributed training is common in deep learning use cases, and I think it would be very helpful to have an example of Optuna in the context of distributed training. If such an example already exists and I have missed it, my apologies, but I would be glad to be directed towards it!

## Description

The following is a rough schema for a possible approach to using Optuna together with TorchElastic.

### Assumptions

The main design assumption for this template is that we are proposing a schema for an Optuna optimization script that can be executed by the [TorchElastic launch script](https://github.com/pytorch/elastic/blob/v0.2.2/torchelastic/distributed/launch.py), and that will execute all Optuna optimization and training within this script.

- Assume a world size of n elastic workers
- All workers should have access to the same trial
- The trial should propose identical hyperparameters per worker
- Only one worker should return optimization results for the trial

### Logic of Example Optimization script

- Initialize a GPU process group across all workers
- One worker (random local rank) executes a `study.optimize` call to obtain hyperparameters before executing training logic
- All workers initialize a temporary process group where the Optuna worker broadcasts its hyperparameters to all other workers
- All workers execute identical training logic
- The Optuna worker returns the optimization metric value on completion
- If any of the workers are interrupted, the process group is restarted and the trial is marked as failed

See [this example](https://github.com/pytorch/elastic/blob/v0.2.2/examples/imagenet/main.py) for a reference regarding the usage of temporary process groups to broadcast data within a worker group.

### Complications

We may need to store additional state across surviving workers about the number of successfully completed trials to date. If the worker group restarts because of e.g. node failure, simply using an optimise call with a fixed `n_trials` can result in more trials than anticipated being run.

## Alternatives

This is a topic that I haven't yet found clear design patterns for. There may be better options than my proposed schema, which it would be great to discuss!

In general the goal is to come up with a clear and functional example of integration between Optuna and TorchElastic, which can possibly also give inspiration for integrating Optuna with other distributed training frameworks.

**Comment by twolffpiggott** [comment]
I realise this issue might be better suited to [optuna-examples](https://github.com/optuna/optuna-examples); happy to move it over if this is the case.

**Comment by Crissman** [comment]
Thanks, @twolffpiggott ! Please do move this over to optuna-examples, and close this issue.

**Comment by Crissman** [comment]
Closing this issue as an example issue. 🙇‍♂️

**Comment by twolffpiggott** [comment]
Thanks @Crissman; apologies for the delay on this; will open in examples.

**Comment by nzw0301** [comment]
@twolffpiggott  could I transfer this issue to the example's issue by using a function of github?

**Comment by twolffpiggott** [comment]
Yup, no problem. I'd actually be happy to contribute when you open the new one. I implemented the proposal and it works well with TorchElastic :)
